rfcs/text/0000-dss-deletion-daemon.md

```diff
@@ -0,0 +1,268 @@
+
+Deletion Table Entry Example:
+
+|Bundle.Version|admin|reason|AWS Deletion Markers (key,ID)|GCP Previous Generations (key, previous generation)|
+|--------------|-----|------|--------------------|------------------------|
+|1234-2134-3145|admin@email.com|consent|(file.obj, 1234), ...| (file.obj, 4321), ...|
+
+A bundle is restored in AWS by deleting the deletion marker and the tombstone for each files in the bundle.
+In GCS a bundle can be restored by by storing a copy of the orginal files in the bucket with the same name, and deleting 
+the tombstones.
+
+### Resources
+
+* [AWS S3 Versioning](https://docs.aws.amazon.com/AmazonS3/latest/dev/Versioning.html),
+* [AWS S3 Deleting Object Versions](https://docs.aws.amazon.com/AmazonS3/latest/dev/DeletingObjectVersions.html),
+* [AWS S3 Restoring Previous Versions](https://docs.aws.amazon.com/AmazonS3/latest/dev/RestoringPreviousVersions.html),
+* [AWS Object Lifecycle Management](https://docs.aws.amazon.com/AmazonS3/latest/dev/object-lifecycle-mgmt.html),
+* [Amazon DynamoDB TTL](https://aws.amazon.com/about-aws/whats-new/2017/02/amazon-dynamodb-now-supports-automatic-item-expiration-with-time-to-live-ttl/)
+* [GCS Object Versioning](https://cloud.google.com/storage/docs/object-versioning),
+* [GCS Object Lifecycle Management](https://cloud.google.com/storage/docs/lifecycle)
+* [HumanCellAtlas/data-store#1662 Explore Bucket Versioning for Deletion](https://github.com/HumanCellAtlas/data-store/issues/1662)
+* [HumanCellAtlas/data-store#1663 Deletion Daemon Design](https://github.com/HumanCellAtlas/data-store/issues/1663)
+
+### Open Issues
+
+TODO: remove support for `all-versions` tombstone in Data Store
+
+TODO: implement consistency check for dangling bundles
+
+TODO: implement object deletion and restoration for AWS anf GCP
+
+TODO: implement delete_bundle.py
+
+TODO: implement restore_bundle.py
+
+TODO: enable bucket versioning and lifecycle rules for replicas.
+
+TODO: Remove `Bundle/Delete` API from DSS.
+
+### Acceptance Criteria [optional]
+
+*Acceptance criteria are the conditions that a RFC must satisfy to be accepted by users or other stakeholders.*
+
+* When a bundle is tombstoned it is made immediately unavailable to users using search.
+* Maintainers of external indices shall removed deleted data from its index when a bundle tombstone notification is received.
+* Bundle files are still available until the deletion process has run.
+* Bundles are not completely removed from the DSS until after the grace period has elapsed.
+
+### Unresolved Questions
+
+- *What aspects of the design do you expect to clarify further through the RFC review process?*
+- *What aspects of the design do you expect to clarify later during iterative development of this RFC?*
+
+### Drawbacks and Limitations [optional]
+
+* This introduces the ability to permanently destroy data from the DSS. 
+* There is no limit on the amount of data that can be rapidly permanently deleted.
```

Comment:
We should put in a rate limiter.

Reply:
Rather than a rate limiter there will be a mandatory grace period before permanent deletion and notifications sent to admins to inform them about pending deletions. The admin will then have a window of time to reverse any or all deletions.

Reply:
probably the best combo:

1. Grace period
1. Metrics for deletion rate + an alert on high deletion rates